rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,530 @@
+the tables, we use `TableID * joint_combiner^max_joint_size` to mix in the
+`TableID`, where `max_joint_size` is the maximum length of any lookup or table
+value. This also simplifies the 'joining' of the `table_id` polynomial
+commitment, since we can straightforwardly scale it by the constant
+`joint_combiner^max_joint_size` to find its contribution to every row in the
+combined table.
+
+### Sampling the `joint_combiner`
+
+Currently, the `joint_combiner` is sampled using randomness that depends on the
+witness. However, where we have runtime or side-loaded tables, a malicious
+prover may be able to select values in those tables that abuse knowledge of
+`joint_combiner` to create collisions.
+
+For example, the prover could create the appearance that there is a value
+`(x, y, z)` in the table with ID 1 by entering a single value `a` into the
+table with ID 2, by computing
+```
+a =   x
+    + y * joint_combiner
+    + z * joint_combiner^2
+    + -1 * joint_combiner^max_joint_size
+```
+so that the combined contribution with its table ID becomes
+```
+a + 2 * joint_combiner^max_joint_size
+```
+and thus matches exactly the combined value from `(x, y, z)` in the table with ID 1:
+```
+x + y * joint_combiner + z * joint_combiner^2 + 1 * joint_combiner^max_joint_size
+```
+
+Thus, we need to ensure that the `joint_combiner` depends on the values in
+runtime or side-loaded tables, so that the values in these tables cannot be
+chosen using knowledge of the `joint_combiner`. To do this, we must use the
+commitments to the values in these tables as part of the source of randomness
+for `joint_combiner` in the prover and verifier.
+
+In particular, this means that we must have a commitment per column for each
+type of table, so that the verifier can confirm the correct construction of the
+combined table.
+
+As usual, we will use the 'transcript' of the proof so far -- including all of
+these components -- as the seed for this randomness.
+
+### Representing the combined fixed table
+
+The fixed table can be constructed at constraint system generation time, by
+extending all of the constituent tables to the same width (i.e. number of
+columns), generating the appropriate table IDs array, and concatenating all of
+the tables. Concretely:
+
+Let `t[id][row][col]` be the `col`th element of the `row`th entry in the `id`th
+fixed table.
+Let `W` be the maximum width of all of the `t[id][row]`s.
+For any `t[id][row]` whose width `w` is less than `W`, pad it to width `W` by
+setting `t[id][row][col] = 0` for all `w < col <= W`.
+Form the combined table by concatenating all of these tables
+`FixedTable = t[0] || t[1] || t[2] || ...`
+and store in `TableID` the table ID that the corresponding row came from.
+Then, for every `output_row`, we have
+```
+FixedTable[output_row][col] = t[id][row][col]`
+TableID[output_row] = id
+where output_row = len(t[0]) + len(t[1]) + ... + len(t[id-1]) + row
+```
```

Comment:
I read this over and over and it still doesn't make sense to me. For example, how do you compute the value of fixedtable[0][0] ? Also, what if t[0] and t[1] don't have the same number of rows/entries?

Reply:
I added markdown diagram in https://github.com/o1-labs/proof-systems/pull/536, you might want to steal this to make this part clearer :o